zerver/views/home.py

```diff
@@ -117,14 +116,8 @@ def home(request: HttpRequest) -> HttpResponse:
     if settings.ROOT_DOMAIN_LANDING_PAGE and subdomain == Realm.SUBDOMAIN_FOR_ROOT_DOMAIN:
         return hello_view(request)
 
-    # TODO: The following logic is a bit hard to read. We save a
-    # database query in the common case by avoiding the call to
-    # `get_valid_realm_from_request` if user hasn't requested
-    # web-public access.
-    if (
-        request.POST.get("prefers_web_public_view") == "true"
-        or request.session.get("prefers_web_public_view")
-    ) and get_valid_realm_from_request(request).allow_web_public_streams_access():
+    realm = get_valid_realm_from_request(request)
+    if realm.allow_web_public_streams_access():
         return web_public_view(home_real)(request)
     return zulip_login_required(home_real)(request)
```

Comment:
OK this change is I think what is causing the test failure. One can reproduce by visiting `http://zulipdev.com:9991`, aka the development environment's homepage when no realm exists. We previously served a redirect to the common /devlogin page (no specific realm), and with this change serve a 404 JSON error:

`{"result":"error","msg":"Invalid subdomain","code":"NONEXISTENT_SUBDOMAIN"}`

I assume the cause being that `web_public_view` doesn't have the same "invalid realm" error handling logic as `zulip_login_required`.

Reply:
OK I think the right thing is this diff:
```
diff --git a/zerver/views/home.py b/zerver/views/home.py
index be9abc2c77..a9fc3b0bc4 100644
--- a/zerver/views/home.py
+++ b/zerver/views/home.py
@@ -8,7 +8,7 @@ from django.shortcuts import redirect, render
 from django.utils.cache import patch_cache_control
 
 from zerver.actions.user_settings import do_change_tos_version
-from zerver.context_processors import get_valid_realm_from_request
+from zerver.context_processors import get_realm_from_request
 from zerver.decorator import web_public_view, zulip_login_required
 from zerver.forms import ToSForm
 from zerver.lib.compatibility import is_outdated_desktop_app, is_unsupported_browser
@@ -116,7 +116,9 @@ def home(request: HttpRequest) -> HttpResponse:
     if settings.ROOT_DOMAIN_LANDING_PAGE and subdomain == Realm.SUBDOMAIN_FOR_ROOT_DOMAIN:
         return hello_view(request)
 
-    realm = get_valid_realm_from_request(request)
+    realm = get_realm_from_request(request)
+    if realm is None:
+        return render(request, "zerver/invalid_realm.html", status=404)
     if realm.allow_web_public_streams_access():
         return web_public_view(home_real)(request)
     return zulip_login_required(home_real)(request)
```

Which will result in this scenario appearing like this:

![image](https://user-images.githubusercontent.com/2746074/165653722-26eda738-0f9a-49a5-98fe-0d7470fa27e9.png)

Reply:
We probably want to refine the page for that case; @alya we can discuss and open an issue, but this will end up being the homepage for self-hosted servers that haven't set up a realm yet.